Content.Shared/_Moffstation/Cards/Systems/PlayingCardsSystem.Transfer.cs

```diff
@@ -0,0 +1,351 @@
+    /// <br/>
+    /// If <paramref name="user"/> is not null, audio and a pickup animation will be played.
+    public List<Entity<PlayingCardComponent>> Take<TStack>(
+        Entity<TStack> stack,
+        Range range,
+        EntityCoordinates destinationCoords,
+        EntityUid? user
+    ) where TStack : PlayingCardStackComponent
+    {
+        var ret = new List<Entity<PlayingCardComponent>>();
+        TransferImpl(
+            CardSourceFrom(stack),
+            Transform(stack).Coordinates,
+            (cards, _) =>
+            {
+                ret.AddRange(cards.Select(cardLike => EnsureSpawnedOrNull(cardLike, destinationCoords))
+                    .OfType<Entity<PlayingCardComponent>>());
+                return (ret.FirstOrNull(), ret.Count);
+            },
+            destinationCoords,
+            range,
+            user
+        );
+        return ret;
+    }
+
+    /// Adds all <paramref name="cards"/> given to <paramref name="stack"/>.
+    /// <br/>
+    /// If <paramref name="user"/> is not null, audio and a pickup animation will be played.
+    public void Add<TStack>(
+        Entity<TStack> stack,
+        IEnumerable<Entity<PlayingCardComponent>> cards,
+        EntityCoordinates cardsCoords,
+        EntityUid? user
+    ) where TStack : PlayingCardStackComponent
+    {
+        TransferImpl(
+            (range, _) => cards.Take(range).Select(it => new CardLike.Entity(it)),
+            cardsCoords,
+            CardSinkFrom(stack),
+            Transform(stack).Coordinates,
+            ..,
+            user
+        );
+    }
+
+
+    /// A source of cards for transferring. Cards in the specified <paramref name="range"/> are taken from their current
+    /// location and yielded in the returned enumerable. The <paramref name="user"/> is also provided so that things
+    /// like predicted audio can be played.
+    private delegate IEnumerable<CardLike> CardSource(Range range, EntityUid? user);
+
+    /// A sink for cards for transferring. The given <paramref name="cards"/> are added to this sink, and the first card
+    /// moved along with the total number of cards moved are returned. The first/total are used for animating the
+    /// transfer. The <paramref name="user"/> is also provided so that things like predicted audio can be played.
+    private delegate (Entity<PlayingCardComponent>? firstCard, int numTotalCards) CardSink(
+        IEnumerable<CardLike> cards,
+        EntityUid? user
+    );
+
+    /// This type represents a Card either as an entity with <see cref="PlayingCardComponent"/> or an unspawned card in
+    /// a <see cref="PlayingCardDeckComponent"/> during transfers.
+    private abstract record CardLike : ISealedInheritance
+    {
+        // Private constructor to seal inheritance.
+        private CardLike() { }
+
+        public sealed record Entity(Entity<PlayingCardComponent> Ent) : CardLike;
+
+        public sealed record Unspawned(PlayingCardInDeck Data) : CardLike;
+    }
```

Comment:
`CardSource` is basically just a function type which takes the requested range and yields cards in an enumerable.
`CardSink` is basically just a function type which consumes the enumerable from a `CardSource` and says "here're the cards I moved".

`CardLike` is the abstraction which is used to deal with real card entities versus unspawned cards.
`CardSource` can give up `CardLike.Unspawned`, but `CardSink` **MUST** instantiate them as needed (eg. hands as `CardSink` will spawn cards)